File: clients/node/src/categories/contractMethods.ts
import { IOneShotClient } from '../types/client.js';
import { EthereumAbi } from '../types/abi.js';
import { NewSolidityStructParam } from '../types/struct.js';
import { FullPrompt } from '../types/contract.js';
import {
  ContractMethod,
  ContractMethodList,
  ContractMethodEstimate,
  ContractMethodTestResult,
  ContractMethodEncodeResult,
  ContractMethodParams,
  ContractMethodStateMutability,
  ERC7702Authorization,
} from '../types/contractMethod.js';
import {
  contractMethodSchema,
  contractMethodListSchema,
  contractMethodEstimateSchema,
  listContractMethodsSchema,
  executeContractMethodSchema,
  executeAsDelegatorContractMethodSchema,
  testContractMethodSchema,
  getContractMethodSchema,
  estimateContractMethodSchema,
  encodeContractMethodSchema,
  readContractMethodSchema,
  createContractMethodSchema,
  importFromABISchema,
  updateContractMethodSchema,
  deleteContractMethodSchema,
  contractSearchSchema,
  fullPromptSchema,
  contractMethodTestResultSchema,
  contractMethodEncodeResultSchema,
  assureContractMethodsFromPromptSchema,
} from '../validation/contractMethod.js';
import { Transaction } from '../types/transaction.js';
import { transactionSchema } from '../validation/transaction.js';

export class ContractMethods {
  constructor(private client: IOneShotClient) {}

  /**
   * Execute a contractMethod
   * @param contractMethodId The ID of the contractMethod to execute
   * @param params Configuration parameters for the contractMethod
   * @param options Optional execution options
   * @returns Promise<Transaction>
   * @throws {ZodError} If the parameters are invalid
   */
  async execute(
    contractMethodId: string,
    params: ContractMethodParams,
    options?: {
      walletId?: string;
      memo?: string;
      authorizationList?: ERC7702Authorization[];
      value?: string;
      contractAddress?: string;
    }
  ): Promise<Transaction> {
    const validatedParams = executeContractMethodSchema.parse({
      contractMethodId,
      params,
      ...options,
    });

    const response = await this.client.request<ContractMethod>(
      'POST',
      `/methods/${validatedParams.contractMethodId}/execute`,
      {
        params: validatedParams.params,
        walletId: validatedParams.walletId,
        memo: validatedParams.memo,
        authorizationList: validatedParams.authorizationList,
        value: validatedParams.value,
        contractAddress: validatedParams.contractAddress,
      }
    );

    return transactionSchema.parse(response);
  }

  /**
   * Execute a contractMethod as a delegator
   * @param contractMethodId The ID of the contractMethod to execute as a delegator
   * @param params Configuration parameters for the contractMethod
   * @param options Optional execution options
   * @returns Promise<Transaction>
   * @throws {ZodError} If the parameters are invalid
   */
  async executeAsDelegator(
    contractMethodId: string,
    delegatorAddress: string,
    params: ContractMethodParams,
    options: {
      walletId?: string;
      memo?: string;
      value?: string;
    }
  ): Promise<Transaction> {
    const validatedParams = executeAsDelegatorContractMethodSchema.parse({
      contractMethodId,
      delegatorAddress,
      params,
      ...options,
    });

    const response = await this.client.request<ContractMethod>(
      'POST',
      `/methods/${validatedParams.contractMethodId}/executeAsDelegator`,
      {
        params: validatedParams.params,
        walletId: validatedParams.walletId,
        memo: validatedParams.memo,
        delegatorAddress: validatedParams.delegatorAddress,
        value: validatedParams.value,
      }
    );

    return transactionSchema.parse(response);
  }

  /**
   * Test a contractMethod without actually executing it
   * @param contractMethodId The ID of the contractMethod to test
   * @param params Configuration parameters for the contractMethod
   * @param options Optional test options
   * @returns Promise<ContractMethodTestResult>
   * @throws {ZodError} If the parameters are invalid
   */
  async test(
    contractMethodId: string,
    params: ContractMethodParams,
    options?: {
      authorizationList?: ERC7702Authorization[];
      value?: string;
      contractAddress?: string;
    }
  ): Promise<ContractMethodTestResult> {
    const validatedParams = testContractMethodSchema.parse({
      contractMethodId,
      params,
      ...options,
    });

    const response = await this.client.request<ContractMethodTestResult>(
      'POST',
      `/methods/${validatedParams.contractMethodId}/test`,
      {
        params: validatedParams.params,
        authorizationList: validatedParams.authorizationList,
        value: validatedParams.value,
        contractAddress: validatedParams.contractAddress,
      }
    );

    return contractMethodTestResultSchema.parse(response);
  }

  /**
   * Get a contractMethod by ID
   * @param id ContractMethod ID
   * @returns Promise<ContractMethod>
   * @throws {ZodError} If the ID is invalid
   */
  async get(id: string): Promise<ContractMethod> {
    const validatedParams = getContractMethodSchema.parse({ id });

    const response = await this.client.request<ContractMethod>(
      'GET',
      `/methods/${validatedParams.id}`
    );

    return contractMethodSchema.parse(response);
  }

  /**
   * List contractMethods for a business
   * @param businessId The business ID to list contractMethods for
   * @param params Optional filter parameters
   * @returns Promise<ContractMethodList>
   * @throws {ZodError} If the parameters are invalid
   */
  async list(
    businessId: string,
    params?: {
      pageSize?: number;
      page?: number;
      chainId?: number;
      name?: string;
      status?: 'live' | 'archived' | 'both';
      contractAddress?: string;
      promptId?: string;
    }
  ): Promise<ContractMethodList> {
    // Validate all parameters using the schema
    const validatedParams = listContractMethodsSchema.parse({
      businessId,
      ...params,
    });

    const queryParams = new URLSearchParams();
    Object.entries(validatedParams).forEach(([key, value]) => {
      if (value !== undefined && key !== 'businessId') {
        queryParams.append(key, value.toString());
      }
    });
    const queryString = queryParams.toString();
    const path = queryString
      ? `/business/${validatedParams.businessId}/methods?${queryString}`
      : `/business/${validatedParams.businessId}/methods`;

    const response = await this.client.request<ContractMethodList>('GET', path);

    // Validate the response
    return contractMethodListSchema.parse(response);
  }

  /**
   * Estimate the cost of executing a contractMethod
   * @param contractMethodId The ID of the contractMethod to estimate
   * @param params Configuration parameters for the contractMethod
   * @param options Optional estimation options
   * @returns Promise<ContractMethodEstimate>
   * @throws {ZodError} If the parameters are invalid
   */
  async estimate(
    contractMethodId: string,
    params: ContractMethodParams,
    options?: {
      authorizationList?: ERC7702Authorization[];
      value?: string;
    }
  ): Promise<ContractMethodEstimate> {
    const validatedParams = estimateContractMethodSchema.parse({
      contractMethodId,
      params,
      ...options,
    });

    const response = await this.client.request<ContractMethodEstimate>(
      'POST',
      `/methods/${validatedParams.contractMethodId}/estimate`,
      {
        params: validatedParams.params,
        authorizationList: validatedParams.authorizationList,
        value: validatedParams.value,
      }
    );

    return contractMethodEstimateSchema.parse(response);
  }

  /**
   * Encode a contractMethod to get the transaction data
   * @param contractMethodId The ID of the contractMethod to encode
   * @param params Configuration parameters for the contractMethod
   * @param options Optional encoding options
   * @returns Promise<ContractMethodEncodeResult>
   * @throws {ZodError} If the parameters are invalid
   */
  async encode(
    contractMethodId: string,
    params: ContractMethodParams,
    options?: {
      authorizationList?: ERC7702Authorization[];
      value?: string;
    }
  ): Promise<ContractMethodEncodeResult> {
    const validatedParams = encodeContractMethodSchema.parse({
      contractMethodId,
      params,
      ...options,
    });

    const response = await this.client.request<ContractMethodEncodeResult>(
      'POST',
      `/methods/${validatedParams.contractMethodId}/encode`,
      {
        params: validatedParams.params,
        authorizationList: validatedParams.authorizationList,
        value: validatedParams.value,
      }
    );

    return contractMethodEncodeResultSchema.parse(response);
  }

  /**
   * Read the result of a view or pure function
   * @param contractMethodId The ID of the contractMethod to read
   * @param params Configuration parameters for the contractMethod
   * @returns Promise<any> The JSON value returned by the function
   * @throws {ZodError} If the parameters are invalid
   */
  async read(contractMethodId: string, params: ContractMethodParams): Promise<any> {
    const validatedParams = readContractMethodSchema.parse({
      contractMethodId,
      params,
    });

    const response = await this.client.request<any>(
      'POST',
      `/methods/${validatedParams.contractMethodId}/read`,
      { params: validatedParams.params }
    );

    return response;
  }

  /**
   * Create a new contractMethod
   * @param businessId The business ID to create the contractMethod for
   * @param params ContractMethod creation parameters
   * @returns Promise<ContractMethod>
   * @throws {ZodError} If the parameters are invalid
   */
  async create(
    businessId: string,
    params: {
      chainId: number;
      contractAddress: string;
      walletId: string;
      name: string;
      description: string;
      functionName: string;
      stateMutability: ContractMethodStateMutability;
      inputs: NewSolidityStructParam[];
      outputs: NewSolidityStructParam[];
      callbackUrl?: string | null;
    }
  ): Promise<ContractMethod> {
    const validatedParams = createContractMethodSchema.parse({
      businessId,
      ...params,
    });

    const response = await this.client.request<ContractMethod>(
      'POST',
      `/business/${validatedParams.businessId}/methods`,
      validatedParams
    );

    return contractMethodSchema.parse(response);
  }

  /**
   * Import contractMethods from an ABI
   * @param businessId The business ID to create the contractMethods for
   * @param params ABI import parameters
   * @returns Promise<ContractMethod[]>
   * @throws {ZodError} If the parameters are invalid
   */
  async importFromABI(
    businessId: string,
    params: {
      chainId: number;
      contractAddress: string;
      walletId: string;
      name: string;
      description: string;
      abi: EthereumAbi;
    }
  ): Promise<ContractMethod[]> {
    const validatedParams = importFromABISchema.parse({
      businessId,
      ...params,
    });

    const response = await this.client.request<ContractMethod[]>(
      'POST',
      `/business/${validatedParams.businessId}/methods/abi`,
      validatedParams
    );

    return response.map((item) => contractMethodSchema.parse(item));
  }

  /**
   * Update a contractMethod
   * @param contractMethodId The ID of the contractMethod to update
   * @param params Update parameters
   * @returns Promise<ContractMethod>
   * @throws {ZodError} If the parameters are invalid
   */
  async update(
    contractMethodId: string,
    params: {
      chainId?: number;
      contractAddress?: string;
      walletId?: string;
      name?: string;
      description?: string;
      functionName?: string;
      stateMutability?: ContractMethodStateMutability;
      callbackUrl?: string | null;
    }
  ): Promise<ContractMethod> {
    const validatedParams = updateContractMethodSchema.parse({
      contractMethodId,
      ...params,
    });

    const response = await this.client.request<ContractMethod>(
      'PUT',
      `/methods/${validatedParams.contractMethodId}`,
      validatedParams
    );

    return contractMethodSchema.parse(response);
  }

  /**
   * Delete a contractMethod
   * @param contractMethodId The ID of the contractMethod to delete
   * @returns Promise<void>
   * @throws {ZodError} If the ID is invalid
   */
  async delete(contractMethodId: string): Promise<void> {
    const validatedParams = deleteContractMethodSchema.parse({
      contractMethodId,
    });

    await this.client.request<void>('DELETE', `/methods/${validatedParams.contractMethodId}`);
  }

  /**
   * Search for contracts
   * @param query The search query
   * @param params Optional search parameters
   * @returns Promise<FullPrompt[]>
   * @throws {ZodError} If the parameters are invalid
   */
  async search(query: string, params?: { chainId?: number }): Promise<FullPrompt[]> {
    const validatedParams = contractSearchSchema.parse({
      query,
      ...params,
    });

    const response = await this.client.request<FullPrompt[]>(
      'POST',
      '/prompts/search',
      validatedParams
    );

    return fullPromptSchema.array().parse(response);
  }

  /**
   * Assure contract methods exist for a given Prompt
   * @param businessId The business ID to create the contract methods for
   * @param params Parameters for creating contract methods from a Prompt
   * @returns Promise<ContractMethod[]>
   * @throws {ZodError} If the parameters are invalid
   */
  async assureContractMethodsFromPrompt(
    businessId: string,
    params: {
      chainId: number;
      contractAddress: string;
      walletId: string;
      promptId?: string;
    }
  ): Promise<ContractMethod[]> {
    const validatedParams = assureContractMethodsFromPromptSchema.parse({
      businessId,
      ...params,
    });

    const response = await this.client.request<ContractMethod[]>(
      'POST',
      `/business/${validatedParams.businessId}/methods/prompt`,
      validatedParams
    );

    return contractMethodSchema.array().parse(response);
  }
}

File: clients/go/contractMethods.go
package oneshot

import (
	"context"
	"encoding/json"

	swagger "github.com/UXlySoftware/1shot-clients/clients/go/gen"
	"github.com/antihax/optional"
)

// ContractMethods handles all contract method-related operations
type ContractMethods struct {
	api        *swagger.ContractMethodsApiService
	businessId string
}

// Execute executes a contract method
func (t *ContractMethods) Execute(
	ctx context.Context,
	contractMethodId string,
	params map[string]interface{},
	walletId, memo *string,
	authorizationList []swagger.Erc7702Authorization,
	value *string,
	contractAddress *string,
) (*swagger.Transaction, error) {
	body := swagger.ContractMethodIdExecuteBody{}
	if params != nil {
		jsonBytes, err := json.Marshal(params)
		if err != nil {
			return nil, err
		}
		var jsonValue swagger.JsonValue
		if err := json.Unmarshal(jsonBytes, &jsonValue); err != nil {
			return nil, err
		}
		body.Params = &jsonValue
	}
	if walletId != nil {
		body.WalletId = *walletId
	}
	if memo != nil {
		body.Memo = *memo
	}
	if authorizationList != nil {
		body.AuthorizationList = authorizationList
	}
	if value != nil {
		body.Value = *value
	}
	if contractAddress != nil {
		body.ContractAddress = *contractAddress
	}
	resp, _, err := t.api.MethodsContractMethodIdExecutePost(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// ExecuteAsDelegator executes a contract method as a delegator
func (t *ContractMethods) ExecuteAsDelegator(
	ctx context.Context,
	contractMethodId string,
	params map[string]interface{},
	walletId, memo *string,
	delegatorAddress string,
) (*swagger.Transaction, error) {
	body := swagger.ContractMethodIdExecuteAsDelegatorBody{
		DelegatorAddress: delegatorAddress,
	}
	if params != nil {
		jsonBytes, err := json.Marshal(params)
		if err != nil {
			return nil, err
		}
		var jsonValue swagger.JsonValue
		if err := json.Unmarshal(jsonBytes, &jsonValue); err != nil {
			return nil, err
		}
		body.Params = &jsonValue
	}
	if walletId != nil {
		body.WalletId = *walletId
	}
	if memo != nil {
		body.Memo = *memo
	}
	resp, _, err := t.api.MethodsContractMethodIdExecuteAsDelegatorPost(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// Test tests a contract method without executing it
func (t *ContractMethods) Test(ctx context.Context, contractMethodId string, params map[string]interface{}) (*swagger.ContractMethodTestResult, error) {
	body := swagger.ContractMethodIdTestBody{}
	if params != nil {
		jsonBytes, err := json.Marshal(params)
		if err != nil {
			return nil, err
		}
		var jsonValue swagger.JsonValue
		if err := json.Unmarshal(jsonBytes, &jsonValue); err != nil {
			return nil, err
		}
		body.Params = &jsonValue
	}
	resp, _, err := t.api.MethodsContractMethodIdTestPost(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// Encode encodes the transaction data for a contract method
func (t *ContractMethods) Encode(ctx context.Context, contractMethodId string, params map[string]interface{}) (*swagger.ContractMethodEncodeResult, error) {
	body := swagger.ContractMethodIdEncodeBody{}
	if params != nil {
		jsonBytes, err := json.Marshal(params)
		if err != nil {
			return nil, err
		}
		var jsonValue swagger.JsonValue
		if err := json.Unmarshal(jsonBytes, &jsonValue); err != nil {
			return nil, err
		}
		body.Params = &jsonValue
	}
	resp, _, err := t.api.MethodsContractMethodIdEncodePost(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// Get gets a contract method by ID
func (t *ContractMethods) Get(ctx context.Context, id string) (*swagger.ContractMethod, error) {
	resp, _, err := t.api.MethodsContractMethodIdGet(ctx, id)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// List lists contract methods for a business
func (t *ContractMethods) List(
	ctx context.Context,
	pageSize, page *int32,
	chainId *swagger.EChain,
	name *string,
	status *swagger.EDeletedStatusSelector,
	contractAddress *string,
	promptId *string,
	methodType *string,
) ([]swagger.ContractMethod, error) {
	opts := &swagger.ContractMethodsApiBusinessBusinessIdMethodsGetOpts{}
	if pageSize != nil {
		opts.PageSize = optional.NewInterface(*pageSize)
	}
	if page != nil {
		opts.Page = optional.NewInterface(*page)
	}
	if chainId != nil {
		opts.ChainId = optional.NewInterface(*chainId)
	}
	if name != nil {
		opts.Name = optional.NewString(*name)
	}
	if status != nil {
		opts.Status = optional.NewInterface(*status)
	}
	if contractAddress != nil {
		opts.ContractAddress = optional.NewInterface(*contractAddress)
	}
	if promptId != nil {
		opts.PromptId = optional.NewInterface(*promptId)
	}
	if methodType != nil {
		opts.MethodType = optional.NewString(*methodType)
	}

	resp, _, err := t.api.BusinessBusinessIdMethodsGet(ctx, t.businessId, opts)
	if err != nil {
		return nil, err
	}
	return resp.Response, nil
}

// Estimate estimates the cost of executing a contract method
func (t *ContractMethods) Estimate(ctx context.Context, contractMethodId string, params map[string]interface{}) (*swagger.ContractMethodEstimate, error) {
	body := swagger.ContractMethodIdEstimateBody{}
	if params != nil {
		jsonBytes, err := json.Marshal(params)
		if err != nil {
			return nil, err
		}
		var jsonValue swagger.JsonValue
		if err := json.Unmarshal(jsonBytes, &jsonValue); err != nil {
			return nil, err
		}
		body.Params = &jsonValue
	}
	resp, _, err := t.api.MethodsContractMethodIdEstimatePost(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// Read reads the result of a view or pure function
func (t *ContractMethods) Read(ctx context.Context, contractMethodId string, params map[string]interface{}) (map[string]interface{}, error) {
	body := swagger.ContractMethodIdReadBody{}
	if params != nil {
		jsonBytes, err := json.Marshal(params)
		if err != nil {
			return nil, err
		}
		var jsonValue swagger.JsonValue
		if err := json.Unmarshal(jsonBytes, &jsonValue); err != nil {
			return nil, err
		}
		body.Params = &jsonValue
	}
	resp, _, err := t.api.MethodsContractMethodIdReadPost(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	jsonBytes, err := json.Marshal(resp)
	if err != nil {
		return nil, err
	}
	var result map[string]interface{}
	if err := json.Unmarshal(jsonBytes, &result); err != nil {
		return nil, err
	}
	return result, nil
}

// Create creates a new contract method
func (t *ContractMethods) Create(
	ctx context.Context,
	chain swagger.EChain,
	contractAddress string,
	walletId string,
	name string,
	description string,
	functionName string,
	stateMutability swagger.ESolidityStateMutability,
	inputs []swagger.NewSolidityStructParam,
	outputs []swagger.NewSolidityStructParam,
	callbackUrl *string,
) (*swagger.ContractMethod, error) {
	body := swagger.BusinessIdMethodsBody{
		ChainId:         &chain,
		ContractAddress: contractAddress,
		WalletId:        walletId,
		Name:            name,
		Description:     description,
		FunctionName:    functionName,
		StateMutability: &stateMutability,
		Inputs:          inputs,
		Outputs:         outputs,
	}
	if callbackUrl != nil {
		body.CallbackUrl = *callbackUrl
	}
	resp, _, err := t.api.BusinessBusinessIdMethodsPost(ctx, body, t.businessId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// ImportFromABI imports contract methods from an ABI
func (t *ContractMethods) ImportFromABI(
	ctx context.Context,
	chain swagger.EChain,
	contractAddress string,
	walletId string,
	name string,
	description string,
	abi []swagger.OneOfEthereumAbiItems,
) ([]swagger.ContractMethod, error) {
	body := swagger.MethodsAbiBody{
		ChainId:         &chain,
		ContractAddress: contractAddress,
		WalletId:        walletId,
		Name:            name,
		Description:     description,
		Abi:             &abi,
	}
	resp, _, err := t.api.BusinessBusinessIdMethodsAbiPost(ctx, body, t.businessId)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

// AssureFromPrompt assures that contract methods exist for a given prompt
func (t *ContractMethods) AssureFromPrompt(
	ctx context.Context,
	chain swagger.EChain,
	contractAddress string,
	walletId string,
	promptId *string,
) ([]swagger.ContractMethod, error) {
	body := swagger.MethodsPromptBody{
		ChainId:         &chain,
		ContractAddress: contractAddress,
		WalletId:        walletId,
	}
	if promptId != nil {
		body.PromptId = *promptId
	}
	resp, _, err := t.api.BusinessBusinessIdMethodsPromptPost(ctx, body, t.businessId)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

// Update updates a contract method
func (t *ContractMethods) Update(
	ctx context.Context,
	contractMethodId string,
	chain *swagger.EChain,
	contractAddress, walletId, name, description, functionName, callbackUrl *string,
) (*swagger.ContractMethod, error) {
	body := swagger.MethodsContractMethodIdBody{}
	if chain != nil {
		body.ChainId = chain
	}
	if contractAddress != nil {
		body.ContractAddress = *contractAddress
	}
	if walletId != nil {
		body.WalletId = *walletId
	}
	if name != nil {
		body.Name = *name
	}
	if description != nil {
		body.Description = *description
	}
	if functionName != nil {
		body.FunctionName = *functionName
	}
	if callbackUrl != nil {
		body.CallbackUrl = *callbackUrl
	}
	resp, _, err := t.api.MethodsContractMethodIdPut(ctx, body, contractMethodId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// Delete deletes a contract method
func (t *ContractMethods) Delete(ctx context.Context, contractMethodId string) error {
	_, err := t.api.MethodsContractMethodIdDelete(ctx, contractMethodId)
	return err
}
